indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/common/OrderedSequenceNumber.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.seekablestream.common;
+
+
+import java.util.Objects;
+
+/**
+ * Represents a Kafka/Kinesis stream sequence number. Mainly used to do
+ * comparison and indicate whether the sequence number is exclusive
+ *
+ * @param <T> type of sequence number
+ */
+public abstract class OrderedSequenceNumber<T> implements Comparable<OrderedSequenceNumber<T>>
+{
+  private final T sequenceNumber;
+  private final boolean isExclusive;
+  private final boolean useExclusive;
+
+  protected OrderedSequenceNumber(T sequenceNumber, boolean useExclusive, boolean isExclusive)
```

Comment:
As mentioned previously, having both `useExclusive` and `isExclusive` is pretty confusing. 

It also might be worthwhile adding comments about what 'exclusive' even means and why it's necessary.